pkg/util/quotapool/int_rate_impl_test.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package quotapool
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+)
+
+// Limit defines a rate in terms of quota per second.
+type Limit float64
+
+// RateLimiter implements a token-bucket style rate limiter.
+// It has the added feature that quota acquired from the pool can be returned
+// in the case that they end up not getting used.
+type RateLimiter struct {
+	qp *QuotaPool
+
+	// TODO(ajwerner): synchronization around changing limits.
+	burst     int64
+	rateLimit Limit
+}
+
+// NewRateLimiter defines a new RateLimiter.
+func NewRateLimiter(name string, limit Limit, burst int64, options ...Option) *RateLimiter {
+	rl := &RateLimiter{rateLimit: limit, burst: burst}
+	bucket := rateBucket{
+		p:           rl,
+		cur:         float64(burst),
+		lastUpdated: timeutil.Now(),
+	}
+	rl.qp = New(name, &bucket, options...)
+	bucket.lastUpdated = rl.qp.timeSource.Now()
+	return rl
+}
+
+// Acquire acquires n quota from the RateLimiter. This acquired quota may be
+// released back into the token bucket or it may be consumed.
+func (rl *RateLimiter) Acquire(ctx context.Context, n int64) (*RateAlloc, error) {
+	if err := rl.WaitN(ctx, n); err != nil {
+		return nil, err
+	}
+	return (*RateAlloc)(rl.newRateAlloc(n)), nil
+}
+
+// WaitN acquires n quota from the RateLimiter. This acquisition cannot be
+// released.
+func (rl *RateLimiter) WaitN(ctx context.Context, n int64) error {
+	if n == 0 {
+		// Special case 0 acquisition.
+		return nil
+	}
+	r := rl.newRateRequest(n)
+	defer rl.putRateRequest(r)
+	if err := rl.qp.Acquire(ctx, r); err != nil {
+		return err
+	}
+	return r.err
+}
+
+// rateBucket is the implementation of Resource which remains in the quotapool
+// for a RateLimiter.
+type rateBucket struct {
+	p           *RateLimiter
+	cur         float64
+	lastUpdated time.Time
+}
+
+var _ Resource = (*rateBucket)(nil)
+
+func (i *rateBucket) Merge(other Resource) {
+	o := other.(*rateAlloc)
+	i.cur += float64(o.alloc)
+	o.rl.putRateAlloc(o)
+
+	if i.cur > float64(i.p.burst) {
+		i.cur = float64(i.p.burst)
+	}
+}
+
+// RateAlloc is an allocated quantity of quota which can be released back into
+// the token-bucket RateLimiter.
+type RateAlloc struct {
+	alloc int64
+	rl    *RateLimiter
+}
+
+// Return returns the RateAlloc to the RateLimiter. It is not safe to call any
```

Comment:
There's no technical problem with later adding code that returns quota to the pool that wasn't initially in it, right? Thinking about Ben's proposal. In that case, we probably also want to keep the quota even if it overflows the bucket (i.e. exceeds burst).

Reply:
(but one step at a time).